cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -0,0 +1,1722 @@
+    if (closed_) {
+      return Status::Invalid("Operation on closed stream");
+    }
+    return pos_;
+  }
+
+  Status Write(const std::shared_ptr<Buffer>& buffer) override {
+    return DoAppend(buffer->data(), buffer->size(), buffer);
+  }
+
+  Status Write(const void* data, int64_t nbytes) override {
+    return DoAppend(data, nbytes);
+  }
+
+  Status DoAppend(const void* data, int64_t nbytes,
+                  std::shared_ptr<Buffer> owned_buffer = nullptr) {
+    if (closed_) {
+      return Status::Invalid("Operation on closed stream");
+    }
+    auto content = Azure::Core::IO::MemoryBodyStream(
+        const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(data)), nbytes);
+    auto result = fileClient_->Append(content, 0);
+    return Status::OK();
+  }
+
+  Status Flush() override {
+    if (closed_) {
+      return Status::Invalid("Operation on closed stream");
+    }
+    fileClient_->Flush(content_length_);
+    return Status::OK();
+  }
+
+ protected:
+  std::shared_ptr<Azure::Storage::Files::DataLake::DataLakePathClient> pathClient_;
+  std::shared_ptr<Azure::Storage::Files::DataLake::DataLakeFileClient> fileClient_;
+  const io::IOContext io_context_;
+  const AzurePath path_;
+
+  bool closed_ = true;
+  int64_t pos_ = 0;
+  int64_t content_length_ = kNoSize;
+  const std::shared_ptr<const KeyValueMetadata> metadata_;
+};
+
+TimePoint ToTimePoint(int secs) {
+  std::chrono::nanoseconds ns_count(static_cast<int64_t>(secs) * 1000000000);
+  return TimePoint(std::chrono::duration_cast<TimePoint::duration>(ns_count));
+}
+
+void FileObjectToInfo(
+    const Azure::Storage::Files::DataLake::Models::PathProperties& properties,
+    FileInfo* info) {
+  info->set_type(FileType::File);
+  info->set_size(static_cast<int64_t>(properties.FileSize));
+  info->set_mtime(ToTimePoint((int)(std::chrono::duration_cast<std::chrono::seconds>(
+                                        properties.LastModified - Azure::DateTime(1970))
+                                        .count())));
+}
+
+void PathInfoToFileInfo(const std::string path, const FileType type, const int64_t size,
+                        const Azure::DateTime dt, FileInfo* info) {
+  info->set_type(type);
+  info->set_size(size);
+  info->set_path(path);
+  info->set_mtime(ToTimePoint(
+      (int)(std::chrono::duration_cast<std::chrono::seconds>(dt - Azure::DateTime(1970))
+                .count())));
+}
+
+}  // namespace
+
+// -----------------------------------------------------------------------
+// Azure filesystem implementation
+
+class AzureBlobFileSystem::Impl
+    : public std::enable_shared_from_this<AzureBlobFileSystem::Impl> {
+ public:
+  io::IOContext io_context_;
+  std::shared_ptr<Azure::Storage::Blobs::BlobServiceClient> gen1Client_;
+  std::shared_ptr<Azure::Storage::Files::DataLake::DataLakeServiceClient> gen2Client_;
+  std::string dfs_endpoint_url;
+  std::string blob_endpoint_url;
+  bool isHierarchicalNamespaceEnabled;
```

Comment:
Style nit: can you suffix member variables with an underscore consistently, and also use snake case?
```suggestion
  std::string dfs_endpoint_url_;
  std::string blob_endpoint_url_;
  bool is_hierarchical_namespace_enabled_;
```